### PR TITLE
8320215: HeapDumper can use DumpWriter buffer during merge

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix updates HeapMerger to use writer buffer (no need to copy memory, also writer buffer is 1MB instead of 4KB).
Additionally fixed small issue in FileWriter (looks like `ssize_t` instead of `size_t` is a typo, the argument should be unsigned)

Testing: all HeapDump-related tests on Oracle supported platforms

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320215](https://bugs.openjdk.org/browse/JDK-8320215): HeapDumper can use DumpWriter buffer during merge (**Enhancement** - P4)

### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Yi Yang](https://openjdk.org/census#yyang) (@y1yang0 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18850/head:pull/18850` \
`$ git checkout pull/18850`

Update a local copy of the PR: \
`$ git checkout pull/18850` \
`$ git pull https://git.openjdk.org/jdk.git pull/18850/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18850`

View PR using the GUI difftool: \
`$ git pr show -t 18850`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18850.diff">https://git.openjdk.org/jdk/pull/18850.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18850#issuecomment-2065519132)

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
Looks good.

**Comment by alexmenkov** [comment]
Ping. Can I get second review please.

File: src/hotspot/share/services/heapDumper.cpp
Comment (plummercj) [inline src/hotspot/share/services/heapDumper.cpp:2137]: Why flush on each iteration instead of just once after you are done with the loop?

**Comment by y1yang0** [comment]
I remember experimenting with different buffer sizes and figuring out that 4KB was the sweet spot. We could potentially switch to 1MB, but it would be better if we had some benchmark numbers to back that up.

File: src/hotspot/share/services/heapDumper.cpp
Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp:2137]: Standard way to use AbstractDumpWriter is to use `AbstractDumpWriter::write_XXX` methods (they all call `write_raw(const void* s, size_t len)` which copies data from the provided memory to `_writer->buffer()`, and flushes when the buffer is full)
This code uses writer internals - on each iteration we read up to 1MB from `segment_fs` directly to `_writer->buffer()`, so we need to flush it before performing next read, otherwise data in the buffer will be overridden.

**Comment by alexmenkov** [comment]
> I remember experimenting with different buffer sizes and figuring out that 4KB was the sweet spot. We could potentially switch to 1MB, but it would be better if we had some benchmark numbers to back that up.

My experiments shown that with the fix time of the merge decreased ~5%. 
But this is not just change of the buffer size:
Before the fix merger (1) read `segment_fs` in 4K chunks, (2) wrote them (copying memory) to `_writer->_buffer` until the buffer is full (i.e. 256 read iterations) and then (3) flush `_writer->_buffer`;
With the fix (1) and (2) are replaced with single reading of 1MB directly to `_writer->_buffer` (no memory copy).

**Review by y1yang0 — APPROVED** [review]
Thanks for the explanation, looks good then~

**Review by plummercj — COMMENTED** [review]
Looks good.

**Pull request closed without merging**
